R/linear_pool_sample.R

```diff
@@ -0,0 +1,128 @@
+#' input model_out_tbl but do preserve relationships across combinations of
+#' task ID variables.
+#'
+#' @importFrom rlang .data
+linear_pool_sample <- function(model_out_tbl, weights = NULL,
+                               weights_col_name = "weight",
+                               model_id = "hub-ensemble",
+                               task_id_cols = NULL,
+                               n_output_samples = NULL) {
+
+  validate_sample_inputs(weights, weights_col_name, n_output_samples)
+
+  num_models <- length(unique(model_out_tbl$model_id))
+  samples_per_model <- model_out_tbl |>
+    dplyr::group_by(dplyr::across("model_id")) |>
+    dplyr::summarize(provided_n_component_samples = dplyr::n()) |>
+    dplyr::ungroup() |>
+    dplyr::select("model_id", "provided_n_component_samples") |>
+    dplyr::distinct(.keep_all = TRUE)
+  unique_provided_samples <- unique(samples_per_model[["provided_n_component_samples"]])
+
+  if (is.null(weights)) {
+    weights <- data.frame(
+      model_id = unique(model_out_tbl$model_id),
+      weight = 1 / num_models,
+      stringsAsFactors = FALSE
+    )
+    weights_col_name <- "weight"
+  }
+  unique_weights <- unique(weights[[weights_col_name]])
+
+  if (length(unique_weights) != 1 || length(unique_provided_samples) != 1 || !is.null(n_output_samples)) {
+    cli::cli_abort(
+      "The requested ensemble calculation doesn't satisfy all conditions:
+      1) {.arg model_out_tbl} contains the same number of samples from each component model,
+      2) {.arg weights} are {.val NULL} or equal for every model,
+      3) {.arg n_output_samples} = {.val NULL}"
+    )
+  }
+
+  model_out_tbl |>
+    make_sample_indices_unique() |>
+    dplyr::select(-"model_id") |>
+    dplyr::mutate(model_id = model_id, .before = 1)
+}
+
+
+#' Make the output type ID values of sample forecasts unique for the same
+#' combination of task IDs but different models
+#'
+#' @param model_out_tbl an object of class `model_out_tbl` with component
+#'   model outputs (e.g., predictions).
+#'
+#' @details The new `output_type_id` column values will follow one of two patterns,
+#' depending on whether the column is detected to be numeric:
+#'   1. If the output type ID is not numeric (may be a character):
+#'      A concatenation of the prediction's model ID and original output type ID
+#'   2. If the output type ID is numeric: A numeric representation of the above
+#'      pattern rendered as a factor.
+#'
+#' @return a model_out_tbl object with unique output type ID values for different
+#'   models but otherwise identical to the input model_out_tbl.
+make_sample_indices_unique <- function(model_out_tbl) {
+  numeric_output_type_ids <- ifelse(is.numeric(model_out_tbl$output_type_id), TRUE, FALSE)
+
+  if (!identical(unique(model_out_tbl$output_type), "sample")) {
+    cli::cli_abort("{.arg model_out_tbl} should only contain the sample output type")
+  }
```

Comment:
How about moving this validation check into `validate_sample_inputs`?  It would be nice to: (a) group validations together as much as possible; and (b) do this check earlier in the process.